bolt-simple-taproot.md

```diff
@@ -0,0 +1,1325 @@
+
+Instead, the responder will simply accept what the initiator proposes. The
+responder can always CPFP after the fact if they require a higher fee rate. The
+initiator is the one that pays fees directly (coming out of their settled
+output), so the responder will always have their full funds develiered to them.
+This change ensures that cooperative close always succeeds after a single
+round.
+
+
+
+### Channel Operation
+
+`n_a_L`: Alice's local secret nonce
+
+`n_B_R`: Bob's remote public nonce
+
+![](taproot_channel_operation.jpg)
+
+#### `commitment_signed` Extensions
+
+A new TLV stream is added to the `commitment_signed` message:
+
+1. `tlv_stream`: `commitment_signed_tlvs`
+2. types:
+   1. type: 2 (`partial_signature_with_nonce`)
+   2. data:
+      * [`98*byte`: `partial_signature || public_nonce`]
+
+##### Requirements
+
+The sender:
+
+  - MUST set the original, non-TLV `signature` field to a 0-byte-array of length
+    64.
+  
+  - MUST retrieve the `musig2` public key previously aggregated from the sorted
+    `funding_pubkey`s using the `KeyAgg` algorithm from `bip-musig2`.
+  
+  - MUST generate a unique nonce to combine with the `next_local_nonce`
+    previously received in the latest of `channel_ready`, `channel_reestablish`,
+    or `revoke_and_ack` using `NonceAgg` from `bip-musig2`.
+  
+  - MUST use the generated secret nonce and the calculated aggregate nonce to
+    construct a `musig2` partial signature for the sender's remote commitment
+    using the `Sign` algorithm from `bip-musig2`.
+  
+  - MUST include the partial signature and the public counterpart of the
+    generated nonce in the `partial_signature_with_nonce` field.
+  
+  - MUST compute each HTLC signature according to
+    [BIP 342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki), as
+    a BIP 340 Schnorr signature (non-partial).
+
+    - Each HTLC signature must be packed as a 64-byte byte value within the
+      existing `htlc_signature` field.
+
+
+The recipient:
+
+  - MUST fail the channel if `signature` is non-zero.
+
+  - MUST fail the channel if `partial_signature_with_nonce` is absent.
+
+  - MUST compute the aggregate nonce from:
+
+    - the `next_local_nonce` field the recipient previously sent in the latest
+      of `channel_ready`, `channel_reestablish`, or `revoke_and_ack`
+
+    - the `public_nonce` included as part of the `partial_signature_with_nonce`
+      field
+
+  - MUST verify the `partial_signature_with_nonce` field using the
+    `PartialSigVerifyInternal` algorithm of `bip-musig2`:
+
+    - if the partial signature is invalid, MUST fail the channel
+
+  - MUST fail the channel if _any_ of the received HTLC signatures is invalid
+    according to BIP 342.
+
+#### `revoke_and_ack` Extensions
+
+A new TLV stream is added to the `revoke_and_ack` message:
+
+1. `tlv_stream`: `revoke_and_ack_tlvs`
+2. types:
+   1. type: 4 (`next_local_nonce`)
+   2. data:
+      * [`66*byte`: `public_nonce`]
```

Comment:
The [counter based nonce generation](https://github.com/lightning/bolts/blob/e25132d8de0164224578964fcd3f7328ddfc3281/bolt-simple-taproot.md#counter-based-nonce-generation--jit-nonce-strengthening) scheme recommends using the commitment index encoded as a 48-bit integer across the sequence and lock time of the commitment transaction to avoid having to store nonces.

Each splice commitment would be at the same commitment state, but have a different funding tx. Does this mean different splice commitments would use the same set of verification public nonces? 

My understanding is that reusing the same nonce and key for different messages would be unsafe. If that is the case, then we might need to rethink the indexing scheme to leave room for unique indexes for each pending splice commitment with the same commitment index.

Reply:
> We still have a single revoke_and_ack message that revokes those N commitments at once, because they still share the same revocation secret.

Interesting, for some reason I thought maybe they'd all have different revocation secrets, so you'd still send N RAA messages, but this makes sense. 

> To be able to generate signatures for the N next commitments, revoke_and_ack must then contain N next_local_nonce, right?

That's right, the signer needs N nonces from the other peer to be able to generate N new commitment sigs for the next state. 

Is there a hard upper limit on the number of pending splices permitted?

Reply:
> Each splice commitment would be at the same commitment state, but have a different funding tx. Does this mean different splice commitments would use the same set of verification public nonces?

Not factoring in the message comment above, yes. IIUC, each splice transaction also has its own protocol-specific sequence number. If this is encoded in the splice kick off transaction, then you can use this as another dimension if generating the nonces with the counter based scheme. So it'd be `sha_chain.leaf(spliceNo || commitNo)`. 

> My understanding is that reusing the same nonce and key for different messages would be unsafe.

Correct! I think w/ the above we can just use the splice sequence number as another dimension for each signed commitment?

Reply:
That sounds like a reasonable approach.

The `commitNo` must already be stored for each commit, but implementers would need to just store the `spliceNo` for the last splice confirmed on-chain until the funding tx of a new splice replaces it.

Reply:
Agreed, there is an ordering on splice transactions, we can use that to generate distinct nonces using the shachain, that shouldn't be an issue.

> Is there a hard upper limit on the number of pending splices permitted?

Not yet, but there should be one, because we cannot send an unbounded number of nonces in `revoke_and_ack` (since message size is limited). In practice the number of pending splices is bounded though, because nodes require the feerate to increase across splices.